storybookjs/storybook
# React: Fix decorators to conditionally render children

Fixes #15223
Custom preview-api hooks assumed that all decorators were always rendered however if a custom decorator conditionally rendered it's children, then not all decorators would get rendered, especially jsxDecorator. This would result in the error
"Rendered more hooks than during the previous render." This removes the assumption that all decorators render every time and relies on each decorator to register itself during MOUNT phase which is handled when each decorator goes through `hookify`

Closes #15223



## What I did

Previously hooks.ts would always set hooks.prevMountedDecorators to the entire list of registered decorators, which assumed they would all render and mount. This changes that behavior to set that to an empty Set in `init` and only update that in `applyDecorators` if it's nullish.
As each decorator goes through the MOUNT stage, then it will add itself to the `prevMountedDecorators` set so it will accurately only move to the UPDATE stage if MOUNT has already happened.

There are a lot of comments at the bottom of #15223 detailing the process I went through to discover this and what I learned and how I came to this conclusion and fix.


## How to test

See repro at https://github.com/marybeshaw/storybook-issue-15223

Started storybook with `yarn storybook` then any story will trigger the error, I used Button mostly.

I couldn't get the instructions for yarn linking to work locally, so I checked out that repo, updated it to yarn 3 and storybook 7.1.0-alpha.0, then manuallly made the changes to node_modules/@storybook/preview/dist/runtime.js to verify the fix. Also used the unit test to confirm the desired behavior. Without the fix, the unit test fails with the "Rendered more hooks" error.

Also verified against repro: https://github.com/seppzero/sb-decorator-issue

Once again, couldn't get linking to work, so I checked out the repo, built it using the 6.5 version it had, ran storybook, the Button story fails with the "Rendered more hooks" error, manually patched node_modules/@storybook/addons/dist/esm/hooks.js , restarted storybook, now the Button story works fine

I don't believe any documentation needs to be updated for this.

## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

      make sure to add the `ci:merged` or `ci:daily` GH label to it.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
